File: src/MbtiEnterpriseSimilarity.App/Services/SimilarityAnalyzer.cs
using MbtiEnterpriseSimilarity.App.Domain;

namespace MbtiEnterpriseSimilarity.App.Services;

public sealed class SimilarityAnalyzer
{
    public StudentProfile FindTarget(IReadOnlyCollection<StudentProfile> profiles, string targetId)
    {
        if (profiles.Count == 0)
        {
            throw new InvalidOperationException("No valid profiles available for analysis.");
        }

        if (string.IsNullOrWhiteSpace(targetId))
        {
            throw new ArgumentException("Target ID is required.", nameof(targetId));
        }

        var target = profiles.FirstOrDefault(p => p.Id.Equals(targetId, StringComparison.OrdinalIgnoreCase));

        return target ?? throw new InvalidOperationException($"Target ID '{targetId}' not found in valid profiles.");
    }

    public StudentProfile FindTarget(
        IReadOnlyCollection<StudentProfile> profiles,
        string targetId,
        IReadOnlyCollection<string> excludedIds)
    {
        var exclusionSet = BuildExclusionSet(excludedIds);
        var normalizedTargetId = targetId.Trim();

        if (exclusionSet.Contains(normalizedTargetId))
        {
            throw new InvalidOperationException($"Target ID '{normalizedTargetId}' is excluded by --exclude-id.");
        }

        var analysisProfiles = ApplyExclusions(profiles, exclusionSet);
        return FindTarget(analysisProfiles, normalizedTargetId);
    }

    public IReadOnlyList<SimilarityMatch> FindNearest(
        IReadOnlyCollection<StudentProfile> profiles,
        string targetId,
        int topN)
        => FindNearest(profiles, targetId, topN, SimilarityMode.Raw, [], DimensionWeights.Equal);

    public IReadOnlyList<SimilarityMatch> FindNearest(
        IReadOnlyCollection<StudentProfile> profiles,
        string targetId,
        int topN,
        SimilarityMode mode,
        IReadOnlyCollection<string> excludedIds)
        => FindNearest(profiles, targetId, topN, mode, excludedIds, DimensionWeights.Equal);

    public IReadOnlyList<SimilarityMatch> FindNearest(
        IReadOnlyCollection<StudentProfile> profiles,
        string targetId,
        int topN,
        SimilarityMode mode,
        IReadOnlyCollection<string> excludedIds,
        DimensionWeights weights)
    {
        if (topN <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(topN), "Top N must be greater than 0.");
        }

        var exclusionSet = BuildExclusionSet(excludedIds);
        var normalizedTargetId = targetId.Trim();

        if (exclusionSet.Contains(normalizedTargetId))
        {
            throw new InvalidOperationException($"Target ID '{normalizedTargetId}' is excluded by --exclude-id.");
        }

        var analysisProfiles = ApplyExclusions(profiles, exclusionSet);
        var target = FindTarget(analysisProfiles, normalizedTargetId);
        var scoreSpace = mode switch
        {
            SimilarityMode.Raw => BuildRawScoreMap(analysisProfiles),
            SimilarityMode.ZScore => BuildZScoreMap(analysisProfiles),
            _ => throw new ArgumentOutOfRangeException(nameof(mode), mode, "Unsupported similarity mode.")
        };

        var nearest = analysisProfiles
            .Where(profile => !profile.Id.Equals(target.Id, StringComparison.OrdinalIgnoreCase))
            .Select(profile =>
            {
                var distance = scoreSpace[target.Id].WeightedEuclideanDistanceTo(scoreSpace[profile.Id], weights);
                var differences = target.Scores.AbsoluteDifferenceFrom(profile.Scores);
                return new SimilarityMatch(profile, distance, differences);
            })
            .OrderBy(match => match.Distance)
            .ThenBy(match => match.Candidate.Id, StringComparer.Ordinal)
            .Take(topN)
            .ToList();

        return nearest;
    }

    private static IReadOnlyDictionary<string, CognitiveScores> BuildRawScoreMap(IReadOnlyCollection<StudentProfile> profiles)
    {
        var map = new Dictionary<string, CognitiveScores>(StringComparer.OrdinalIgnoreCase);

        foreach (var profile in profiles)
        {
            if (!map.TryAdd(profile.Id, profile.Scores))
            {
                throw new InvalidOperationException($"Duplicate ID found in dataset: '{profile.Id}'.");
            }
        }

        return map;
    }

    private static IReadOnlyDictionary<string, CognitiveScores> BuildZScoreMap(IReadOnlyCollection<StudentProfile> profiles)
    {
        var profileList = profiles.ToList();

        if (profileList.Count < 2)
        {
            throw new InvalidOperationException("At least 2 valid profiles are required for z-score normalization.");
        }

        var matrix = profileList
            .Select(profile => ToVector(profile.Scores))
            .ToList();

        var mean = new double[8];
        var stddev = new double[8];

        for (var dim = 0; dim < 8; dim++)
        {
            mean[dim] = matrix.Select(vector => vector[dim]).Average();
            var variance = matrix
                .Select(vector => Math.Pow(vector[dim] - mean[dim], 2))
                .Average();
            stddev[dim] = Math.Sqrt(variance);
        }

        var zScoreMap = new Dictionary<string, CognitiveScores>(StringComparer.OrdinalIgnoreCase);

        for (var i = 0; i < profileList.Count; i++)
        {
            var source = matrix[i];
            var z = new double[8];

            for (var dim = 0; dim < 8; dim++)
            {
                z[dim] = stddev[dim] > 1e-12
                    ? (source[dim] - mean[dim]) / stddev[dim]
                    : 0d;
            }

            if (!zScoreMap.TryAdd(profileList[i].Id, FromVector(z)))
            {
                throw new InvalidOperationException($"Duplicate ID found in dataset: '{profileList[i].Id}'.");
            }
        }

        return zScoreMap;
    }

    private static double[] ToVector(CognitiveScores scores) =>
    [
        scores.Ne,
        scores.Ni,
        scores.Te,
        scores.Ti,
        scores.Se,
        scores.Si,
        scores.Fe,
        scores.Fi
    ];

    private static CognitiveScores FromVector(IReadOnlyList<double> vector) => new(
        vector[0],
        vector[1],
        vector[2],
        vector[3],
        vector[4],
        vector[5],
        vector[6],
        vector[7]);

    private static IReadOnlyList<StudentProfile> ApplyExclusions(
        IReadOnlyCollection<StudentProfile> profiles,
        ISet<string> exclusionSet)
    {
        if (profiles.Count == 0)
        {
            throw new InvalidOperationException("No valid profiles available for analysis.");
        }

        var filtered = profiles
            .Where(profile => !exclusionSet.Contains(profile.Id))
            .ToList();

        if (filtered.Count == 0)
        {
            throw new InvalidOperationException("No profiles remain after applying --exclude-id filters.");
        }

        return filtered;
    }

    private static HashSet<string> BuildExclusionSet(IReadOnlyCollection<string> excludedIds)
    {
        var set = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var id in excludedIds)
        {
            if (!string.IsNullOrWhiteSpace(id))
            {
                set.Add(id.Trim());
            }
        }

        return set;
    }
}

File: src/main/java/com/mbti/similarity/service/SimilarityAnalyzer.java
package com.mbti.similarity.service;

import com.mbti.similarity.domain.CognitiveScores;
import com.mbti.similarity.domain.DimensionWeights;
import com.mbti.similarity.domain.SimilarityMatch;
import com.mbti.similarity.domain.SimilarityMode;
import com.mbti.similarity.domain.StudentProfile;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

public final class SimilarityAnalyzer {
    public StudentProfile findTarget(Collection<StudentProfile> profiles, String targetId) {
        List<StudentProfile> profileList = List.copyOf(Objects.requireNonNull(profiles, "profiles"));
        if (profileList.isEmpty()) {
            throw new IllegalStateException("No valid profiles available for analysis.");
        }

        String normalizedTargetId = normalizeId(targetId, "Target ID is required.");
        for (StudentProfile profile : profileList) {
            if (profile.getId().equalsIgnoreCase(normalizedTargetId)) {
                return profile;
            }
        }

        throw new IllegalStateException("Target ID '" + normalizedTargetId + "' not found in valid profiles.");
    }

    public StudentProfile findTarget(Collection<StudentProfile> profiles, String targetId, Collection<String> excludedIds) {
        Set<String> exclusionSet = buildExclusionSet(excludedIds);
        String normalizedTargetId = normalizeId(targetId, "Target ID is required.");
        if (exclusionSet.contains(normalizedTargetId.toLowerCase(Locale.ROOT))) {
            throw new IllegalStateException("Target ID '" + normalizedTargetId + "' is excluded by --exclude-id.");
        }

        return findTarget(applyExclusions(profiles, exclusionSet), normalizedTargetId);
    }

    public List<SimilarityMatch> findNearest(
        Collection<StudentProfile> profiles,
        String targetId,
        int topN,
        SimilarityMode mode,
        Collection<String> excludedIds,
        DimensionWeights weights
    ) {
        if (topN <= 0) {
            throw new IllegalArgumentException("Top N must be greater than 0.");
        }

        Set<String> exclusionSet = buildExclusionSet(excludedIds);
        String normalizedTargetId = normalizeId(targetId, "Target ID is required.");
        if (exclusionSet.contains(normalizedTargetId.toLowerCase(Locale.ROOT))) {
            throw new IllegalStateException("Target ID '" + normalizedTargetId + "' is excluded by --exclude-id.");
        }

        List<StudentProfile> analysisProfiles = applyExclusions(profiles, exclusionSet);
        StudentProfile target = findTarget(analysisProfiles, normalizedTargetId);
        Map<String, CognitiveScores> scoreSpace = switch (Objects.requireNonNull(mode, "mode")) {
            case RAW -> buildRawScoreMap(analysisProfiles);
            case ZSCORE -> buildZScoreMap(analysisProfiles);
        };

        List<SimilarityMatch> matches = new ArrayList<>();
        for (StudentProfile profile : analysisProfiles) {
            if (profile.getId().equalsIgnoreCase(target.getId())) {
                continue;
            }

            double distance = scoreSpace.get(target.getId().toLowerCase(Locale.ROOT))
                .weightedEuclideanDistanceTo(scoreSpace.get(profile.getId().toLowerCase(Locale.ROOT)), weights);

            matches.add(new SimilarityMatch(
                profile,
                distance,
                target.getScores().absoluteDifferenceFrom(profile.getScores())
            ));
        }

        matches.sort(
            Comparator.comparingDouble(SimilarityMatch::getDistance)
                .thenComparing(match -> match.getCandidate().getId())
        );

        int limit = Math.min(topN, matches.size());
        return List.copyOf(matches.subList(0, limit));
    }

    private static Map<String, CognitiveScores> buildRawScoreMap(List<StudentProfile> profiles) {
        Map<String, CognitiveScores> scoresById = new LinkedHashMap<>();
        for (StudentProfile profile : profiles) {
            String key = profile.getId().toLowerCase(Locale.ROOT);
            if (scoresById.putIfAbsent(key, profile.getScores()) != null) {
                throw new IllegalStateException("Duplicate ID found in dataset: '" + profile.getId() + "'.");
            }
        }
        return scoresById;
    }

    private static Map<String, CognitiveScores> buildZScoreMap(List<StudentProfile> profiles) {
        if (profiles.size() < 2) {
            throw new IllegalStateException("At least 2 valid profiles are required for z-score normalization.");
        }

        List<double[]> matrix = new ArrayList<>(profiles.size());
        for (StudentProfile profile : profiles) {
            matrix.add(profile.getScores().toVector());
        }

        double[] mean = new double[CognitiveScores.DIMENSIONS.size()];
        double[] stddev = new double[CognitiveScores.DIMENSIONS.size()];

        for (int dimension = 0; dimension < CognitiveScores.DIMENSIONS.size(); dimension++) {
            double sum = 0d;
            for (double[] vector : matrix) {
                sum += vector[dimension];
            }
            mean[dimension] = sum / matrix.size();

            double varianceSum = 0d;
            for (double[] vector : matrix) {
                double diff = vector[dimension] - mean[dimension];
                varianceSum += diff * diff;
            }
            stddev[dimension] = Math.sqrt(varianceSum / matrix.size());
        }

        Map<String, CognitiveScores> zScoreMap = new LinkedHashMap<>();
        for (int rowIndex = 0; rowIndex < profiles.size(); rowIndex++) {
            double[] source = matrix.get(rowIndex);
            double[] z = new double[CognitiveScores.DIMENSIONS.size()];
            for (int dimension = 0; dimension < CognitiveScores.DIMENSIONS.size(); dimension++) {
                z[dimension] = stddev[dimension] > 1e-12
                    ? (source[dimension] - mean[dimension]) / stddev[dimension]
                    : 0d;
            }

            StudentProfile profile = profiles.get(rowIndex);
            String key = profile.getId().toLowerCase(Locale.ROOT);
            if (zScoreMap.putIfAbsent(key, CognitiveScores.fromVector(z)) != null) {
                throw new IllegalStateException("Duplicate ID found in dataset: '" + profile.getId() + "'.");
            }
        }

        return zScoreMap;
    }

    private static List<StudentProfile> applyExclusions(Collection<StudentProfile> profiles, Set<String> exclusionSet) {
        List<StudentProfile> profileList = List.copyOf(Objects.requireNonNull(profiles, "profiles"));
        if (profileList.isEmpty()) {
            throw new IllegalStateException("No valid profiles available for analysis.");
        }

        List<StudentProfile> filtered = new ArrayList<>();
        for (StudentProfile profile : profileList) {
            if (!exclusionSet.contains(profile.getId().toLowerCase(Locale.ROOT))) {
                filtered.add(profile);
            }
        }

        if (filtered.isEmpty()) {
            throw new IllegalStateException("No profiles remain after applying --exclude-id filters.");
        }

        return filtered;
    }

    private static Set<String> buildExclusionSet(Collection<String> excludedIds) {
        Set<String> exclusionSet = new LinkedHashSet<>();
        if (excludedIds == null) {
            return exclusionSet;
        }

        for (String id : excludedIds) {
            if (id != null && !id.trim().isEmpty()) {
                exclusionSet.add(id.trim().toLowerCase(Locale.ROOT));
            }
        }

        return exclusionSet;
    }

    private static String normalizeId(String id, String message) {
        if (id == null || id.trim().isEmpty()) {
            throw new IllegalArgumentException(message);
        }

        return id.trim();
    }
}
